### PR TITLE
feat(app-check): added support for firebase app check

### LINKED ISSUE #693: Does this package support Firebase app check?
For verifying firebase app check token?

[https://firebase.google.com/docs/app-check/custom-resource-backend](https://firebase.google.com/docs/app-check/custom-resource-backend)

### PR DESCRIPTION
**Pull request by nhaynes**
# Description

I added support for **Firebase App Check** so the SDK can be used to implement a [custom provider](https://firebase.google.com/docs/app-check/custom-provider) and can also be used to [verify app check tokens](https://firebase.google.com/docs/app-check/custom-resource-backend). 

I wanted to use App Check to protect custom backend resources for a project of mine and was frustrated to find out that it's only available in the official Node.JS SDK.

Closes [#693](https://github.com/kreait/firebase-php/issues/693)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by jeromegamez** [comment]
Wow, this looks like you have put a lot of work into this, and I appreciate that you stuck with the code style of the SDK (for better or worse, it's an old code base, after all, 😅), and even docs and tests! That's the best PR this project has received so far, I love it!

It will not be long before I drop support for PHP <8.1 (I changed my support policy to only support actively supported PHP versions), so this feature will certainly be added to the upcoming 7.0 release. I'm giving you this heads-up because it means a rebase will be in your future 😅.

This could also be a good start for deprecating https://github.com/kreait/firebase-tokens-php and incorporating Custom Token generation and ID Token verification using `firebase/php-jwt` into the SDK, but that's another topic (that could also lead to another rebase).

I appreciate what you did already - should I makle a lot of change requests and comments (I only quickly skimmed the code, so I don't know if there will be many 😄), please keep in mind that I only have the best intentions and appreciate the time and work you're investing 🙏 

I'm pretty swamped at the moment, but I'll try to look at the details as soon as possible! In the meantime, I'll enable running the workflows, in case it already brings something to light that you want/need to address 💪

**Comment by jeromegamez** [comment]
Ha, the tests are already failing due to the PHP requirements. Could you change the requirement in the composer.json to `"php": "~8.1.0|~8.2.0"`, then do a `composer bump` while we're at it, and update the requirements in the `.github/workflows` directory accordingly?

**Comment by nhaynes** [comment]
@jeromegamez Thanks for the feedback! Adding it to 7.x seems like a sensible plan to me. I'll work on getting the php requirements sorted out for now.

**Comment by jeromegamez** [comment]
@nhaynes I will create a new 7.x branch with the updated PHP requirements so that you can re-target the PR.

I will now have a closer look at the changes, but from what I can see, I can already say that I would prefer not to have the phpfastcache dependency (= implementation) - there are already the `withVerifierCache()` and `withAuthTokenCache()` methods in the factory, and I see the App Check Token caching either using the same cache as provided to `withVerifierCache()` or an additional `withAppCheckTokenVerifierCache()`.

This way, we wouldn't have an additional dependency, and the project implementing the SDK can inject its own cache implementation - for example, in Laravel https://github.com/kreait/laravel-firebase/blob/14812145bbd8598ec2c9f447aac166b451cfadd6/src/FirebaseProjectManager.php#L94-L105

File: composer.json
Comment (jeromegamez) [inline composer.json]: See https://github.com/kreait/firebase-php/pull/747#issuecomment-1304133629 for why I would like to have phpfastcache removed again 😅 

Could you instead please add `firebase/php-jwt` as a direct dependency? At the moment it's transitively included through the Google Auth library, but if they were to remove it, we wouldn't have it anymore either.

File: src/Firebase/AppCheck.php
Comment (jeromegamez) [inline src/Firebase/AppCheck.php]: Just a note, you can resolve it instantly :D - it will be so satisfying to replace this with once support for PHP <8.1 is dropped 🤤 

```php
final class AppCheck implements Contract\AppCheck
{
    public function (
        private readonly ApiClient $client,
        private AppCheckTokenGenerator $tokenGenerator,
        private AppCheckTokenVerifier $tokenVerifier,
) {
    // …
}
```

File: src/Firebase/AppCheck/AppCheckTokenOptions.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/AppCheckTokenOptions.php]:
```suggestion
     * @param AppCheckTokenOptionsShape $data
```

File: src/Firebase/AppCheck/AppCheckTokenOptions.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/AppCheckTokenOptions.php]: `ttl` instead of `app_id` in the error message

I'm thinking about removing many of these runtime checks, not just here but in many places of the SDK where PHPStan/Psalm PHPDocs are in place.

For example, here, the `AppCheckTokenOptionsShape` already defines that a `ttl` key _must_ be present, so I'm wondering if runtime checks are even necessary.

So, unless you have strong opinions against it, I think it would be good to remove runtime checks where static analysis would already catch the problem.

(Strong Opinions against it: except "But most people don't use Static Analysis Tools" - because they should™ 😅)

File: src/Firebase/AppCheck/AppCheckTokenVerifier.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/AppCheckTokenVerifier.php]:
```suggestion
    /**
     * @param non-empty-string $projectId
     */
    public function __construct(string $projectId, CachedKeySet $keySet)
```

Annotations like this in general could be added or improved, but don't invest too much time into it, just when you are working at those locations anyway 🙏

File: src/Firebase/AppCheck/DecodedAppCheckToken.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/DecodedAppCheckToken.php]:
```suggestion
     * @param DecodedAppCheckTokenShape $data
```

This would remove the need for the `array_key_exists` checks - it's a public function, but in reality, this will only be used by the SDK itself.

File: src/Firebase/AppCheck/AppCheckTokenVerifier.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/AppCheckTokenVerifier.php:63]: I'm not 100% sure, but I think we could do a

```php
/** @var DecodedAppCheckTokenShape $payload */
$payload = (array) JWT::decode($token, $this->keySet);
```

here, together with my comment on the `DecodedAppCheckToken` this should save us some code 🤞

File: src/Firebase/Factory.php
Comment (jeromegamez) [inline src/Firebase/Factory.php]: This is what I meant when I said that I'd prefer to use an injected instance of a cache instead of declaring it here ourselves - a file cache is undoubtedly good enough for most installations with a filesystem, but there are environments and/or preferences that would require another implementation

**Review by jeromegamez — CHANGES_REQUESTED** [review]
The PR looks very good overall, everything I noted (except phpfastcache) are just minor things that don't really make a change in functionality, so if this is too cumbersome (and you already invested quite some time), I can do it later.

I will create the new 7.x branch with the updated PHP requirements within the hour so that you can retarget the PR (as far as I know this can be done without creating a new PR, but I haven't done that myself yet 😅)

After the next iteration, I'll check out the code on my computer/IDE, that's when I usually find what I missed earlier. :D

**Comment by jeromegamez** [comment]
https://github.com/kreait/firebase-php/tree/7.x is the new default branch 🥳

**Comment by jeromegamez** [comment]
I updated the `beste/*` dependencies to support PHP 8.1 and 8.2, so they should not be the reason for failing tests anymore. Also: #748  - I don't know what the best approach would be here, but we'll figure something out 🤞

**Comment by jeromegamez** [comment]
Sorry for the comment spam and for the back and forth 😅.

Once you've removed the PHPFastCache dependency, and a short review in my IDE, I'll merge the PR.

Language Level migrations and such is positively none of your business 😄

**Comment by jeromegamez** [comment]
#748 is now merged into the 7.x branch.

File: src/Firebase/AppCheck/AppCheckTokenVerifier.php
Comment (nhaynes) [reply src/Firebase/AppCheck/AppCheckTokenVerifier.php:63]: I'm not sure I follow this one. Are you saying the type-hint is all we need here and we shouldn't concern ourselves with the exception handling? I'm not sure I understand how static analysis is going to help us if the developer provides a string that isn't a valid token at runtime.

Comment (nhaynes) [reply src/Firebase/AppCheck/AppCheckTokenVerifier.php:63]: Nevermind, I see what you mean now. You weren't referring to the exception handling, you referring to the `fromArray` call and the type hint there. We are assuming that if the JWT parses correctly, then the fields we expect to be there should be there.

**Comment by jeromegamez** [comment]
Thank you for addressing the things I wanted to have changed, really appreciate it!

The integration tests in this branch run without errors, but I just noticed they do because the Secrets are not available in PR GitHub actions, so they're all skipped.

On my local computer, the unit tests all succeed, but the Integration Tests don't, probably because I don't have a valid AppCheck token

```
GuzzleHttp\Exception\ClientException: Client error: `POST https://firebaseappcheck.googleapis.com/v1/projects/apps/com.example.test-app` resulted in a `404 Not Found` response
```

As I haven't worked with AppCheck Tokens myself yet, could you tell me how to generate one / what I have to do to be able to generate one? The Firebase project I test with has no associated apps, and unfortunately, I'm also not an app developer, so I wouldn't know on my own how to do it 😅 

Other than that, the PR looks good (as the initial one, just better :D), some minor things that I can do myself just before or after merging it.

There's one other thing concerning the AppCheckTokenOptions, but I will comment on that in the code.

I feel like I'm saying "Thank you" too often, but it's heartfelt so I will keep doing it :D

File: src/Firebase/AppCheck/AppCheckTokenOptions.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/AppCheckTokenOptions.php:34]: I don't know why PHPStan didn't catch this, but the AppCheckTokenOptionsShape defines `ttl` to be `string|null`, but here, the check is done with `int`s.

Locally on my computer this gives me 

```
2) Kreait\Firebase\Tests\Integration\AppCheckTest::testCreateTokenWithCustomTtl
TypeError: Kreait\Firebase\AppCheck\AppCheckTokenOptions::__construct(): Argument #1 ($ttl) must be of type ?string, int given, called in /Users/jg/Code/kreait/firebase-php/src/Firebase/AppCheck/AppCheckTokenOptions.php on line 47
```

Not sure what the best approach here could be. I know the Firebase API expects it as a string, so perhaps the object could accept `int`, but expose `string`?

File: src/Firebase/AppCheck/AppCheckTokenOptions.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/AppCheckTokenOptions.php]: Could you please check if the Options are actually JSON-serialized somewhere and/or the `toArray()` method is used anywhere? If not, or if callers could be rewritten to not use json_decode/toArray(), we could perhaps remove them.

(I added those and `__toString()` methods almost everywhere in the beginning and now I regret it because I have to keep them for BC, but also test them :D)

File: src/Firebase/AppCheck/AppCheckTokenVerifier.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/AppCheckTokenVerifier.php]:
```suggestion
     * Verifies the format and signature of a Firebase App Check token.
```

When I start to get nitpicky, it's usually a good sign 😂

File: src/Firebase/AppCheck/AppCheckTokenGenerator.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/AppCheckTokenGenerator.php:44]: Is 300 a configurable value? How does it relate to the `ttl`? Could it be `$exp = $options->ttl ?? 300` or something?

File: src/Firebase/AppCheck/DecodedAppCheckToken.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/DecodedAppCheckToken.php]: Since we have PHP 8.1 as the minimum version, we could go with

```php
private function __construct(
    /** @var non-empty-string $app_id */
    public readonly string $app_id,
    /** @var array<string> $aud /*
    public readonly array $aud,
    public readonly string $exp,
    // and so on
) {
}
```

This way, we could get rid of the property declarations above and the getter functions below 🚀

**Review by jeromegamez — CHANGES_REQUESTED** [review]
I noticed a few more things - as before, nothing serious, but for your consideration 🤞

**Comment by jeromegamez** [comment]
I found out about AppCheck debug tokens and will try it out in the next few days. Once I get it working, I'll merge your PR and get it ready for the new SDK version (i've not decided yet if it will be 7.0 or 7.1, but it will come 😊)

File: src/Firebase/AppCheck/ApiClient.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/ApiClient.php]: This request results in a 404 - according to the [documentation](https://firebase.google.com/docs/reference/appcheck/rest/v1/projects.apps/exchangeCustomToken#http-request), the target URL should be `apps/<appId>:exchangeCustomToken`, how did you make this work in your tests?

File: src/Firebase/Factory.php
Comment (jeromegamez) [inline src/Firebase/Factory.php]:
```suggestion
            'base_uri' => 'https://firebaseappcheck.googleapis.com/v1/projects/'.$projectId.'/',
```

The base URI must end with a slash. Otherwise, a

```php
$http->get('GET', 'app/id')
```

will request `https://firebaseappcheck.googleapis.com/v1/projects/app/id` (which it did for me when I debugged the PR)

File: src/Firebase/Factory.php
Comment (jeromegamez) [inline src/Firebase/Factory.php]: Why is this a link to the Firebase documentation? 😅 

```suggestion
            'https://firebaseappcheck.googleapis.com/v1/jwks',
```

File: tests/Integration/AppCheckTest.php
Comment (jeromegamez) [inline tests/Integration/AppCheckTest.php]: How did you make this work when running the integration tests?

File: tests/Integration/AppCheckTest.php
Comment (jeromegamez) [inline tests/Integration/AppCheckTest.php]: The return types don't need to be checked, the `AppCheckToken` class already ensures that. Same in other tests

**Review by jeromegamez — CHANGES_REQUESTED** [review]
I should have tested locally first instead of just skimming over the Github Web page 😅

File: src/Firebase/AppCheck/AppCheckTokenGenerator.php
Comment (nhaynes) [reply src/Firebase/AppCheck/AppCheckTokenGenerator.php:44]: I based this off of what the official node package is doing. They are always setting the expiration for the custom token to 300. 

I think because it creates a custom token and then immediately exchanges the custom token for a App Check token, that the time here doesn't matter much. I'm guessing they chose 5 mins because it's long enough to ensure a race condition doesn't occur here with the token exchange but it's short enough that it is a short lived token?

File: tests/Integration/AppCheckTest.php
Comment (nhaynes) [reply tests/Integration/AppCheckTest.php]: Apparently I didn't have them running. I didn't realize the integration tests were being auto-skipped locally.

I've now created a test project and added the necessary `test_credentials.json` and `test_rtdb.json` files and some of the integration tests are passing but I'm not sure what I need to put in `test_devices.json` to get the rest passing? What exactly are the registration tokens here?

Comment (jeromegamez) [reply tests/Integration/AppCheckTest.php]: If you run only the AppCheck tests (`phpunit tests/Integration/AppCheckTest.php`), you might not need the FCM registration tokens. I generate one with https://github.com/firebase/quickstart-js/tree/master/messaging and then copy it as a JSON array into the `test_devices.json` file (`["<device token>"]`).

That's a good reminder to add a `CONTRIBUTING.md` 😅

**Comment by jeromegamez** [comment]
@nhaynes I think it should be possible to prevent most of the PHPStan errors when you run `composer install-tools` and then `composer tests`. Also, could you re-request a review (in the right sidebar at the top) once you pushed something, so that I can get a notification and don't have to let you wait? 🙏 

Just because I noticed the 

```php
public function __construct(
        private ServiceAccount $serviceAccount,
        private ClockInterface $clock = SystemClock::create(),
) {
}
```

File: src/Firebase/AppCheck/AppCheckTokenGenerator.php
Comment (jeromegamez) [reply src/Firebase/AppCheck/AppCheckTokenGenerator.php:44]: Sounds reasonable, thanks for clarifying! 🙌

File: src/Firebase/AppCheck/ApiClient.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/ApiClient.php]: I think this would be better located in the `AppCheckToken` class.

File: src/Firebase/AppCheck/AppCheckToken.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/AppCheckToken.php]: I know I added `toArray()` and `jsonSerialize()` to value objects in the past as well. Still, I think as long as the `AppCheckToken` is not serialized in another SDK component, we should omit them here. If a project want to serialize the app check tokens, they can do it themselves 😅

File: src/Firebase/AppCheck/AppCheckTokenGenerator.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/AppCheckTokenGenerator.php]:
```suggestion
    
    private ClockInterface $clock;

    public function __construct(
        private ServiceAccount $serviceAccount,
        ?ClockInterface $clock = null,
    ) {
        $this->clock = $clock ?? SystemClock::create();
    }
```

File: src/Firebase/AppCheck/VerifyAppCheckTokenResponse.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/VerifyAppCheckTokenResponse.php]:
```suggestion
     * @param non-empty-string $appId
```

File: src/Firebase/AppCheck/VerifyAppCheckTokenResponse.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/VerifyAppCheckTokenResponse.php]:
```suggestion
    public function __construct(
        public readonly string $appId,
        public readonly DecodedAppCheckToken $token,
    ) {
    }
```

This would make the `appId()` and `token()` getter functions obsolete. 

When using `private/public <type> $variable` as constructor parameter, you don't need to set them in the constructor body anymore.

When a property is only set in the constructor, we can always add the `readonly` qualifier 💪

File: src/Firebase/AppCheck/DecodedAppCheckToken.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/DecodedAppCheckToken.php]:
```suggestion
    private function __construct(
        public readonly string $app_id,
        /** @var array<string> $aud */
        public readonly array $aud,
        public readonly string $exp,
        public readonly string $iat,
        public readonly string $iss,
        public readonly string $sub,
    ) {
    }
```

Like the others above, this would help us get rid of all the related getter methods.

File: src/Firebase/AppCheck/AppCheckTokenOptions.php
Comment (jeromegamez) [inline src/Firebase/AppCheck/AppCheckTokenOptions.php]:
```suggestion
 *     ttl?: int|null,
```

**Review by jeromegamez — CHANGES_REQUESTED** [review]

**Comment by nhaynes** [comment]
@jeromegamez I believe I've made all the improvements you requested above and I have successfully gotten the **Integration tests** for **AppCheck** to succeed locally (along with **Unit** and **PHPStan**). 

I started a `CONTRIBUTING.md` file to document the steps I had to take to get the integration tests to run locally. I managed to get most of them passing locally following these steps but I am still getting a few errors on the **Auth**, **DynamicLink**, and **RemoveConfig** modules. I'll dig into those later and update `CONTRIBUTING.md` if I find a solution.

**Comment by nhaynes** [comment]
@jeromegamez I added the `guzzlehttp/psr7` library as a direct dependency to ensure the `HttpFactory` interface is available, should address those failed tests.

**Comment by nhaynes** [comment]
@jeromegamez looks like that resolved the test issues and now it's just the code coverage blocking us. I'll add some additional tests this weekend.

**Comment by jeromegamez** [comment]
@nhaynes But don't spend too much time on it, you've already invested so much of it!

I'll probably get to a final review on the week-end.

**Comment by jeromegamez** [comment]
What baffles me a little is that the psr7 library had to be added directly - it should come with Guzzle via the Google auth library. But it doesn't hurt either, I'm just curious.

File: docs/app-check.rst
Comment (jeromegamez) [inline docs/app-check.rst:3]:
```suggestion
########
App Check
########
```

File: docs/app-check.rst
Comment (jeromegamez) [inline docs/app-check.rst:15]:
```suggestion
************************************
Initializing the App Check component
************************************
```

File: docs/app-check.rst
Comment (jeromegamez) [inline docs/app-check.rst:62]: I would document this the same way as, for example, the ID Token verification. Depending on the project, people might retrieve the AppCheck Token in different ways, `getallheaders()` is an implementation details which some users might interpret as "the thing to do" and then open issues when things don't work out 😅 

```suggestion
    $appCheckTokenString = '...';

    try {
        $appCheck->verifyToken($appCheckTokenString);
    } catch (FailedToVerifyAppCheckToken $e) {
```

File: docs/app-check.rst
Comment (jeromegamez) [inline docs/app-check.rst:9]:
```suggestion
* `Introduction to Firebase App Check <https://firebase.google.com/docs/app-check>`_
```

File: docs/app-check.rst
Comment (jeromegamez) [inline docs/app-check.rst:10]:
```suggestion
* `Verify App Check tokens from a custom backend (Client-side) <https://firebase.google.com/docs/app-check/custom-resource-backend>`_
```

File: CHANGELOG-7.x.md
Comment (jeromegamez) [inline CHANGELOG-7.x.md:10]: I wouldn't add dependency changes to the changelog. The firebase-jwt library was already included via the Google Auth Library anyway 😅 
```suggestion
```

File: CHANGELOG-7.x.md
Comment (jeromegamez) [inline CHANGELOG-7.x.md:13]:
```suggestion
* Added support for generating and verifying Firebase App Check Tokens ([#747](https://github.com/kreait/firebase-php/issues/747))
```

File: CONTRIBUTING.md
Comment (jeromegamez) [inline CONTRIBUTING.md:1]:
```suggestion
# Contributing
```

**Review by jeromegamez — CHANGES_REQUESTED** [review]
These are only doc-related changes, and I think we're ready then.

Could you also add `app-check` to the [docs/index.rst](https://github.com/kreait/firebase-php/blob/7.x/docs/index.rst) in the user guide section below the `user-management` entry?

**Comment by jeromegamez** [comment]
Please don't work on the PR anymore, I already have it ready to be merged locally. I will first release a 7.0 and then a 7.1 with the AppCheck component shortly after 🤞 

Thank you so much for your work, I really, really appreciate it - and a lot of other SDK users will do the same, it's been brought up again and again :D 

If you haven't provided this in the context of a job, could you please enable GitHub sponsors or give me a link to a Buy me a coffee or something similar? I think you have definitely earned a token of gratitude 🙏

**Comment by jeromegamez** [comment]
Thank you for your patience! I merged your PR with d2dd8d5a43ee179ab397982567234b39f5f478bd, ed58b866a428684df9cda73b110bf4ae7beaddeb, and 105a128a858e4ee6bf5b8cd72e7f19f8e8c14f1f - the changes I made  were only to adapt to the changes in the other parts of the SDK, but it's still all attributed to you ☺️ 

Thanks again for your work! It will be included in the 7.0 release, which shouldn't take too long. 🤞

**Comment by nhaynes** [comment]
@jeromegamez awesome! thanks for finishing it up!

**Comment by jeromegamez** [comment]
@nhaynes The component has been released with version 7.0 of the SDK. Again, thank you so much for adding it! 🙏 🌺

**Comment by nhaynes** [comment]
@jeromegamez awesome! Great to see! Thanks for getting it across the finish line!

**Pull request closed without merging**
